pkgs/development/python-modules/pytorch/default.nix

```diff
@@ -73,38 +143,83 @@ in buildPythonPackage rec {
   PYTORCH_BUILD_VERSION = version;
   PYTORCH_BUILD_NUMBER = 0;
 
+  BUILD_NAMEDTENSOR = buildNamedTensor;  # experimental feature
+  USE_SYSTEM_NCCL=true;                  # don't build pytorch's third_party NCCL
+
   # Suppress a weird warning in mkl-dnn, part of ideep in pytorch
   # (upstream seems to have fixed this in the wrong place?)
   # https://github.com/intel/mkl-dnn/commit/8134d346cdb7fe1695a2aa55771071d455fae0bc
-  NIX_CFLAGS_COMPILE = lib.optionals (numpy.blasImplementation == "mkl") [ "-Wno-error=array-bounds" ];
+  # https://github.com/pytorch/pytorch/issues/22346
+  #
+  # Also of interest: pytorch ignores CXXFLAGS uses CFLAGS for both C and C++:
+  # https://github.com/pytorch/pytorch/blob/v1.2.0/setup.py#L17
+  NIX_CFLAGS_COMPILE = lib.optionals (my_numpy.blasImplementation == "mkl") [ "-Wno-error=array-bounds" ];
 
   nativeBuildInputs = [
-     cmake
-     utillinux
-     which
+    cmake
+    utillinux
+    which
+    ninja
   ] ++ lib.optionals cudaSupport [ cudatoolkit_joined ];
 
   buildInputs = [
-     numpy.blas
-  ] ++ lib.optionals cudaSupport [ cudnn ]
+    my_numpy.blas
+  ] ++ lib.optionals cudaSupport [ cudnn my_magma nccl ]
     ++ lib.optionals stdenv.isLinux [ numactl ];
 
   propagatedBuildInputs = [
     cffi
-    numpy
+    my_numpy
```

Comment:
Generally you want to avoid propagating different versions of the same library, there might be a scenario where a user will have a pytorch + pyproj environment in which they both use numpy. The numpy dependency from both of those packages will be put on PYTHONPATH, and you can't guarentee the order in which they get traversed. So you may end up in a scenario where you dont have mkl support.

If this was an application, you could override the python package set, and then your dependencies should inherit the overrided packages transitively.

Reply:
@FRidh correct me if I'm wrong, because I want to be :)

Reply:
@stites are you just trying to work around the fact tensorflow has not yet picked up the ability to compile with MKL? (#69454)

Reply:
This was a pattern introduced prior to this PR -- I definitely feel this pain when working on personal projects. Before replicating this pattern in both magma (for mkl) and openMPI (for cuda) I recall reviewing a PR where someone said this was the right course of action. I'd be happy to change it, though, as it did cause some headaches in the beginning.

Ideally, I'd like to trigger a warning that says something like "building with mklSupport, but magma and/or numpy have not been built with mkl." I'm not sure how to do this, though... Maybe just a loud echo in the build phase? openMPISupport can be added to the assert statements.

Reply:
yeah, sorry I introduced this pattern as a newbie crutch :grimacing:. I think the better place to do this is with overrides on `callPackage`? E.g. https://github.com/NixOS/nixpkgs/blob/8cd5aa8a82367eeedbd0261f6d040a106e2a844a/pkgs/top-level/all-packages.nix#L7565

Reply:
the main problem is that when this gets built by hydra, you will just get a cache hit, and pull down the closure. You won't see any build output.

Reply:
cross-referencing python scoping issues #67422

Reply:
Rebuilding pytorch-world with changes now. I'm going to assert that the same mkl/cudatoolkit is sync'd across all dependencies (note the enforced mkl instead of the verbose build output). This mimics how the derivation works now, but requires the user to supply correct dependencies. 

Asserts look something like:
```
assert !cudaSupport || lib.lists.any (inp: inp == cudatoolkit) magma.buildInputs;
```